### PR TITLE
Reset temporal unit parameter value if it becomes no longer valid after temporal unit list changes

### PR DESCRIPTION
**Pull request by ranquild**
Epic https://github.com/metabase/metabase/issues/42118
Demo https://www.loom.com/share/c736e1ffa6a14565b990b9ed295457c5

How to verify:
- Add a question with a breakout based on a date column to a dashboard
- Add a temporal unit parameter and map it to the breakout column
- Set the parameter value e.g. to "Year"
- Modify the list of allowed temporal units for this parameter to exclude "Year"
- The value of the parameter should be reset
- As a side note, the parameter value will be reset when toggling All/None in the temporal unit picker

### CODE REVIEW

File: frontend/src/metabase/dashboard/actions/parameters.unit.spec.ts [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+
+import {
+  REMOVE_PARAMETER,
+  removeParameter,
+  setParameterTemporalUnits,
+} from "./parameters";
+
+function setup(initialState: State) {
+  // eslint-disable-next-line react-hooks/rules-of-hooks
+  const browserHistory = useRouterHistory(createHistory)();
```
Comment (ranquild) [inline frontend/src/metabase/dashboard/actions/parameters.unit.spec.ts:28]: It's not the best but this is how the store in configured in `ui.tsx`

Comment (ranquild) [reply frontend/src/metabase/dashboard/actions/parameters.unit.spec.ts:28]: Btw this is not a hook

File: frontend/src/metabase/dashboard/actions/parameters.unit.spec.ts [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+        default: expectedDefaultValue,
+      });
+      expect(getParameterValues(store.getState())).toMatchObject({
+        [parameterId]: expectedValue,
+      });
+    },
+  );
+});
+
+describe("removeParameter", () => {
```
Comment (ranquild) [inline frontend/src/metabase/dashboard/actions/parameters.unit.spec.ts:150]: existing test

File: frontend/src/metabase-types/store/dashboard.ts [hunk 3]
```diff
@@ -98,7 +99,7 @@ export interface DashboardState {
   dashcards: Record<DashCardId, StoreDashcard>;
   dashcardData: DashCardDataMap;
 
-  parameterValues: Record<ParameterId, ParameterValueOrArray>;
+  parameterValues: ParameterValuesMap;
```
Comment (ranquild) [inline frontend/src/metabase-types/store/dashboard.ts:102]: `ParameterValuesMap` allows `null` values while the previous type does not (incorrect)

File: frontend/src/metabase-lib/v1/urls.ts [hunk 4]
```diff
@@ -5,7 +5,7 @@ import type { ParameterWithTarget } from "metabase-lib/v1/parameters/types";
 import { getParameterValuesBySlug } from "metabase-lib/v1/parameters/utils/parameter-values";
 import { remapParameterValuesToTemplateTags } from "metabase-lib/v1/parameters/utils/template-tags";
 import { isTransientId } from "metabase-lib/v1/queries/utils/card";
-import type { ParameterId, ParameterValueOrArray } from "metabase-types/api";
+import type { ParameterValuesMap } from "metabase-types/api";
```
Comment (kamilmielnik) [inline frontend/src/metabase-lib/v1/urls.ts:8]: :x: It's possible to use a disabled unit of time via URL

https://github.com/metabase/metabase/assets/6830683/9c8a88fb-a2c7-4a02-b21b-8677b539532d

File: frontend/src/metabase-types/store/dashboard.ts (hunk 3 above)
Comment (kamilmielnik) [reply frontend/src/metabase-types/store/dashboard.ts:102]: There are 2 other places in the project that use `Record<ParameterId, ParameterValueOrArray>` - should we update them too?

**Review by kamilmielnik — CHANGES_REQUESTED** [review]

**Pull request closed without merging**